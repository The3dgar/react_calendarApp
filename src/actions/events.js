import Swal from "sweetalert2";
import { fetchWithToken } from "../helper/fetch";
import { prepareEvents } from "../helper/prepareEvents";
import { types } from "../types/types";

// disparar si se graba en la base de datos
export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const resp = await fetchWithToken("/events", event, "POST");
      const body = await resp.json();

      if (body.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};
const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken("events");
      const body = await resp.json();
      console.log(body);
      const events = prepareEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken(`/events/${event.id}`, event, "PUT");
      const body = await resp.json();
      if (body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventStartDelete = (event) => {
  return async (dispatch, getState) => {
    const { activeEvent: event } = getState().calendar;
    try {
      const resp = await fetchWithToken(`/events/${event.id}`, event, "DELETE");
      const body = await resp.json();
      if (body.ok) {
        dispatch(eventDeleted(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventLogout = () => ({
  type: types.eventLogout
})
