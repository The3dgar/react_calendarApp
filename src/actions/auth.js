import Swal from "sweetalert2";
import { fetchWithOutToken, fetchWithToken } from "../helper/fetch";
import { types } from "../types/types";
import { eventLogout } from "./events";

export const startLogin = (email, password) => {
  // dispatch gracias al thunk
  return async (dispatch) => {
    const resp = await fetchWithOutToken("auth", { email, password }, "POST");
    const body = await resp.json();

    if (body.ok) {
      const { token, uid, name } = body;
      localStorage.setItem("token", token);
      localStorage.setItem("token_init", new Date().getTime());

      dispatch(login({ uid, name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const resp = await fetchWithOutToken(
      "auth/new",
      { email, password, name },
      "POST"
    );
    const body = await resp.json();

    if (body.ok) {
      const { token, uid, name } = body;
      localStorage.setItem("token", token);
      localStorage.setItem("token_init", new Date().getTime());

      dispatch(login({ uid, name }));
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startLogout = () => {
  return dispatch => {
    localStorage.clear()
    dispatch(eventLogout())
    dispatch(logout())
  }
}

const logout = () => ({type: types.authLogout})

export const startChecking  = ()  => {
  return async dispatch => {
    const token = localStorage.getItem("token")

    if(!token) return dispatch(checkingFinish())
    
    const resp = await fetchWithToken("auth/renew");
    const body = await resp.json();
    if (body.ok) {
      const { token, uid, name } = body;
      localStorage.setItem("token", token);
      localStorage.setItem("token_init", new Date().getTime());

      dispatch(login({ uid, name }));
    } else {
      Swal.fire("Errores", body.msg, "error");
      dispatch(checkingFinish())
    }
  }
}

const checkingFinish = () => ({
  type: types.authCheckingFinish
})