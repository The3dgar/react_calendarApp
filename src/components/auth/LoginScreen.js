import React from "react";
import "./login.css";
import { useForm } from "../../hooks/useForm";
import { useDispatch } from "react-redux";
import { startLogin, startRegister } from "../../actions/auth";
import Swal from "sweetalert2";
export const LoginScreen = () => {
  const [formLoginValues, handleLoginInputChange] = useForm({
    lEmail: "edgarolivar16@gmail.cl",
    lPassword: "123456",
  });
  const [formRegisterValues, handleRegisterInputChange] = useForm({
    rName: "Moises",
    rEmail: "moisesolivar16@gmail.cl",
    rPassword: "123456",
    rPassword2: "123456",
  });

  const { lEmail, lPassword } = formLoginValues;
  const { rName, rEmail, rPassword, rPassword2 } = formRegisterValues;

  const dispatch = useDispatch();
  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLogin(lEmail, lPassword));
  };

  const handleRegister = e => {
    e.preventDefault();

    if(rPassword !== rPassword2) return Swal.fire("Warning", "Passwords dont match", "warning")
    dispatch(startRegister(rEmail, rPassword, rName))

  }

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={handleLogin}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Correo"
                name="lEmail"
                value={lEmail}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="lPassword"
                value={lPassword}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Nombre"
                name="rName"
                value={rName}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Correo"
                name="rEmail"
                value={rEmail}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="rPassword"
                value={rPassword}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repita la contraseña"
                name="rPassword2"
                value={rPassword2}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Register" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
