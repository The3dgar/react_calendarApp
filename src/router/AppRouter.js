import React, { useEffect } from "react";
import { BrowserRouter as Router, Switch } from "react-router-dom";
import { CalendarScreen } from "../components/calendar/CalendarScreen";
import { LoginScreen } from "../components/auth/LoginScreen";
import { useDispatch, useSelector } from "react-redux";
import { startChecking } from "../actions/auth";
import { PublicRoute } from "./PublicRoute";
import { PrivateRoute } from "./PrivateRoute";

export const AppRouter = () => {

  const dispatch = useDispatch()
  const {checking, uid} = useSelector(state => state.auth)

  useEffect(() => {
      dispatch(startChecking())
  }, [dispatch])

  if(checking) {
    return  <h5>Wait... </h5>
  }

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            exact
            path="/login"
            component={LoginScreen}
            isAuthenticated={ !!uid  }
          />

          <PrivateRoute
            exact
            path= "/"
            component={CalendarScreen}
            isAuthenticated={ !!uid  }
          />
        </Switch>
      </div>
    </Router>
  );
};
