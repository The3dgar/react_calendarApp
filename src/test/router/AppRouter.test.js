import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from "redux-mock-store"
import thunk from "redux-thunk"
import "@testing-library/jest-dom"
import { AppRouter } from '../../router/AppRouter';
import { startChecking } from '../../actions/auth';
import { act } from 'react-dom/test-utils';

jest.mock( "../../actions/auth", () => ({
  startChecking: jest.fn()
}))

const middlewares  = [thunk]
const mockStore = configureStore(middlewares)


describe('testing in <AppRouter/>', () => {
  test('should show <h1>Wait....</h1>', () => {
    const initState = {
      auth: {
        checking: true
      }
    }
    const store = mockStore(initState)
    store.dispatch = jest.fn()
    
    const wrapper = mount(
      <Provider store={store}>
        <AppRouter/>
      </Provider>
    ) 

    expect(wrapper).toMatchSnapshot()
    
  })  

  test('should show public route', () => {
    const initState = {
      auth: {
        checking: false,
        uid: null
      }
    }
    const store = mockStore(initState)
    store.dispatch = jest.fn()

    const wrapper = mount(
      <Provider store={store}>
        <AppRouter/>
      </Provider>
    ) 

    // expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(".login-container").exists()).toBe(true)
    
  })

  test('should show private route', () => {
    const initState = {
      auth: {
        checking: false,
        uid: "123",
        name: "Edgar"
      },
      calendar: {
        events: []
      },
      ui: {
        modalOpen: false
      }
    }
    const store = mockStore(initState)
    store.dispatch = jest.fn()

    const wrapper = mount(
      <Provider store={store}>
        <AppRouter/>
      </Provider>
    ) 

    // expect(wrapper).toMatchSnapshot()
    expect(wrapper.find(".calendar-screen").exists()).toBe(true)
    
  })

  test('should dispatch startChecking', () => {


    const initState = {
      auth: {
        checking: true
      }
    }
    const store = mockStore(initState)
    store.dispatch = jest.fn()
    act(()=> {
      
      const wrapper = mount(
        <Provider store={store}>
          <AppRouter/>
        </Provider>
      ) 

    })
    

    expect(startChecking).toHaveBeenCalled()
  })

})
