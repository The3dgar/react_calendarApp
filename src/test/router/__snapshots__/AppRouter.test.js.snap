// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing in <AppRouter/> should show <h1>Wait....</h1> 1`] = `
<h5>
  Wait... 
</h5>
`;
